### DISCUSSION TITLE
Make the defaultDialect and most others async compatible

### DISCUSSION THREAD

**Discussion by devkral** [sqlalchemy/sqlalchemy#11685] [Ideas]
Make the defaultDialect and most others async compatible

How?

Use a wrapper which translates the sync stuff in async threads.

Is it hard?

No. I explored already a primitive one:

https://github.com/dymmond/databasez/blob/035be9b8d1d330ee528d5c5f9ebff5bbfbb4a182/databasez/utils.py

And usage is in:

https://github.com/dymmond/databasez/blob/035be9b8d1d330ee528d5c5f9ebff5bbfbb4a182/databasez/dialects/jdbc.py

https://github.com/dymmond/databasez/blob/035be9b8d1d330ee528d5c5f9ebff5bbfbb4a182/databasez/dialects/dbapi2.py

Given I am not an sqlalchemy dev with experience my solution can be suboptimal.

**Comment by CaselIT** [comment]
Hi,

That's basically how the aio-x dialect are implemented, but are sadly not too maintained.
Implementing a generic async driver via thread to avoid the aio-x driver was considered, but since most of the interested for async seems to be for postgresql that has async driver noting came of it.

Overall performance will not be great, that's why using thread is not more widely used

**Reply by devkral** [reply]
The idea is to have a thread sensitive way:
every connect starts a thread with the connection.

But improvements are necessary. I am not quite sure if just 1 thread is used for all connections.

**Reply by CaselIT** [reply]
that's what aio-x behaves. IIRC connections are thread subclasses. See for example 
https://github.com/sqlalchemy/sqlalchemy/blob/bae75fe92f9636bafb75461ff0bc556432831e30/lib/sqlalchemy/dialects/sqlite/aiosqlite.py#L202-L203

**Reply by devkral** [reply]
thanks again! This are details I lacked.

Anyway: I meant my implementation. I expect that the dbapi is capable of handling multiple connections in multiple threads. If one thread keeps its connection.

If a dbapi is not thread capable, the pool size of the engine can be set to 1
